### PR TITLE
MatlabCompat update for Julia 1.1.1

### PR DESCRIPTION
**Pull request by aminya**
FemtoCleaner ran to update deprecations

First, I ran my forked FemtoCleaner (https://github.com/aminya/FemtoCleaner.jl)) locally to update the deprecated syntax.

However, as I have written in here in Julia forum (https://discourse.julialang.org/t/add-and-develope-an-unregistered-local-package/25632), I can not add the package, apparently, because of the old Tk.jl image package used. Or maybe there is an issue somewhere else.

When I try to add it I get this error:
```jl
]add C:\\Users\\username\\Documents\\GitHub\\MatlabCompat.jl

  Updating registry at `C:\Users\username\.julia\registries\General`
  Updating git-repo `https://github.com/JuliaRegistries/General.git`
   Cloning git-repo `C:\\Users\\username\\Documents\\GitHub\\MatlabCompat.jl`
  Updating git-repo `C:\\Users\\username\\Documents\\GitHub\\MatlabCompat.jl`
 Resolving package versions...
ERROR: Unsatisfiable requirements detected for package Tk [4225ea8a]:
 Tk [4225ea8a] log:
 ├─possible versions are: [0.0.1, 0.2.0-0.2.17, 0.3.0-0.3.8, 0.4.0, 0.5.0-0.5.2] or uninstalled
 ├─restricted to versions 0.0.0-* by MatlabCompat [54e369d6], leaving only versions [0.0.1, 0.2.0-0.2.17, 0.3.0-0.3.8, 0.4.0, 0.5.0-0.5.2]
 │ └─MatlabCompat [54e369d6] log:
 │   ├─possible versions are: 0.1.2 or uninstalled
 │   └─MatlabCompat [54e369d6] is fixed to version 0.1.2+
 └─restricted by julia compatibility requirements to versions: uninstalled — no versions lef
```
Now in Julia 1.1.1, there are better image packages we can use. Otherwise, We should use https://github.com/JuliaGraphics/Gtk.jl instead.

### CODE REVIEW

File: src/MatlabCompat/imagetools.jl [hunk 1]
```diff
@@ -1,261 +1,257 @@
-module ImageTools
```
Comment (aminya) [inline src/MatlabCompat/imagetools.jl]: Differences:
https://www.diffchecker.com/NLI1oFQi

File: test/runtests.jl [hunk 2]
```diff
@@ -1,106 +1,102 @@
-using Base
```
Comment (aminya) [inline test/runtests.jl]: https://www.diffchecker.com/9MLcxcqs

**Comment by aminya** [comment]
However, when I try to add it I get this error:
```jl
]add C:\\Users\\username\\Documents\\GitHub\\MatlabCompat.jl
  Updating registry at `C:\Users\username\.julia\registries\General`
  Updating git-repo `https://github.com/JuliaRegistries/General.git`
   Cloning git-repo `C:\\Users\\username\\Documents\\GitHub\\MatlabCompat.jl`
  Updating git-repo `C:\\Users\\username\\Documents\\GitHub\\MatlabCompat.jl`
 Resolving package versions...
ERROR: Unsatisfiable requirements detected for package Tk [4225ea8a]:
 Tk [4225ea8a] log:
 ├─possible versions are: [0.0.1, 0.2.0-0.2.17, 0.3.0-0.3.8, 0.4.0, 0.5.0-0.5.2] or uninstalled
 ├─restricted to versions 0.0.0-* by MatlabCompat [54e369d6], leaving only versions [0.0.1, 0.2.0-0.2.17, 0.3.0-0.3.8, 0.4.0, 0.5.0-0.5.2]
 │ └─MatlabCompat [54e369d6] log:
 │   ├─possible versions are: 0.1.2 or uninstalled
 │   └─MatlabCompat [54e369d6] is fixed to version 0.1.2+
 └─restricted by julia compatibility requirements to versions: uninstalled — no versions lef
```jl

**Comment by aminya** [comment]
@ayakimovich fixed the problem for adding my version of MatlabCompat (under development one) by updating the REQUIRE file (commented Tk).

To add this for developing, clone it to some path.
Then use this command:
``]dev pathTopackage``

For example for me it is:
``]dev C:\\Users\\yahyaaba\\Documents\\GitHub\\MatlabCompat.jl``

Now we can run test.jl to fix the old and deprecated syntax. I found some of them here: https://github.com/JuliaImages/ImageCore.jl/commit/e808091ba0f774a443e39526329a31665b4d83b0#diff-18d62f68c87f15977a1239fc3ca4ed58
I added an issue to get some help from someone that knows Images package very well
https://github.com/JuliaImages/ImageCore.jl/issues/83

**Comment by aminya** [comment]
Undergoing now: Choosing between 4 syntaxes that work with the current syntax. Discussion here:
https://github.com/JuliaImages/ImageCore.jl/issues/83

@ayakimovich This is becoming very complex. Because it is not just some API, and types and element type have changed as well.

In the old module, [here](https://github.com/MatlabCompat/MatlabCompat.jl/blob/5c63ce3029fd372fb63d07d77c74cfae1c377ab1/test/runtests.jl#L27), a grayscale image is defined using grayim and its properties are set:
```jl
# make a test image
array = [0.1 0.001 0 0; 0.1 0.401 0 0; 0.1 0.01 0 0; 0.1 0.901 0 0.6];
img = grayim(array);
properties(img)["IMcs"] = "Gray";
```

[Here in my new module](https://github.com/aminya/MatlabCompat.jl/blob/a18f029f2383aa5cd552f040031d5c1472087c8e/test/runtests.jl#L40) I am trying to define a grayscale image from an array. I have come up with 4 different syntaxes, and I am not sure which one is more common or I should choose,
```jl
array = [0.1 0.001 0 0; 0.1 0.401 0 0; 0.1 0.01 0 0; 0.1 0.901 0 0.6];

# Syntax 1 - Array{Gray}
img1 =Gray.(array);
# calculates a grayscale representation of a color image using the Rec 601 luma. (https://juliaimages.org/latest/function_reference/#Color-conversion-1)
imgMet=ImageMeta(img, time="after noon");
println(typeof(img1)) # gives Array{Gray{Float64},2}

# Syntax 2 - ReinterpretArray
img2 = colorview(Gray, array);
# returns a view of the numeric array A, interpreting successive elements of A as if they were channels of Colorant C. (https://juliaimages.org/latest/function_reference/#ImageCore.colorview)
imgMet=ImageMeta(img2, time="after noon"); # to add ImageMeta
println(typeof(img2)) # gives Base.ReinterpretArray{Gray{Float64},2,Float64,Array{Float64,2}}

# Syntax 3 - ReinterpretArray + ImageMeta
imgMet = ImageMeta(colorview(Gray, array), time="after noon")
# returns an ImageMeta type, which holds both data() with Colorant element type, and properties() (Dict type)
properties(imgMet)["location"] = "university"; # to access and extract properties
img3=data(imgMet); # to access and extract data
println(typeof(img3)) # gives Base.ReinterpretArray{Gray{Float64},2,Float64,Array{Float64,2}}

# Syntax 4 - Array + ImageMeta
# this does not work with this module
imgMet = ImageMeta(array, time="after noon")
# returns an ImageMeta type, which holds both data() but array has Float type instead of Colorant, and properties() (Dict type)
properties(imgMet)["location"] = "university"; # to access and extract properties
img4=data(imgMet); # to access and extract data
println(typeof(img4)) # gives Array{Float64,2}
# this synatx does

img=img1; # choose img syntax
```

With three first syntaxes ``eltype(img) <: Gray`` gives [true](https://github.com/aminya/MatlabCompat.jl/blob/a18f029f2383aa5cd552f040031d5c1472087c8e/src/MatlabCompat/imagetools.jl#L61), but the 4th one is just an array.


Also, replacing ``raw(img)`` simply with ``rawview(img)``[ does not work](https://github.com/aminya/MatlabCompat.jl/blob/a18f029f2383aa5cd552f040031d5c1472087c8e/src/MatlabCompat/imagetools.jl#L67) because of no matching methods.

**Comment by johnnychen94** [comment]
Hi, if I understand it right, `imagetools.jl` intends to mimic the MATLAB ImageProcessingToolbox APIs? If that's the case, I guess the best way to develop it is to

1) if it's missing, contribute the real implementation under [JuliaImages](https://github.com/JuliaImages)
2) wrap the real implementation with MATLAB-like APIs, unless there's a severe performance issue.

Just had a glance at the codebase here, I think you can refer to [ImageBinarization](https://github.com/zygmuntszpak/ImageBinarization.jl) when you're updating `graythresh` and `im2bw` 😄

**Comment by johnnychen94** [comment]
> replacing raw(img) simply with rawview(img) does not work because of no matching methods

```julia
help?> rawview
search: rawview

  rawview(img::AbstractArray{FixedPoint})


  returns a "view" of img where the values are interpreted in terms of their raw underlying storage. For example, if
  img is an Array{N0f8}, the view will act like an Array{UInt8}.
```
so `rawview(channelview(img))` would work

I'm not sure if I understand it correctly, but if `imagetools.jl` is treated as another API specification module, I think there are fewer merits to return/use a "raw" numerical array by default (although MATLAB does it this way) compared to `AbstractArray{<:Colorant}`.

**Comment by aminya** [comment]
> 
> 
> Hi, if I understand it right, `imagetools.jl` intends to mimic the MATLAB ImageProcessingToolbox APIs? If that's the case, I guess the best way to develop it is to
> 
>     1. if it's missing, contribute the real implementation under [JuliaImages](https://github.com/JuliaImages)
> 
>     2. wrap the real implementation with MATLAB-like APIs, unless there's a severe performance issue.
> 
> 
> Just had a glance at the codebase here, I think you can refer to [ImageBinarization](https://github.com/zygmuntszpak/ImageBinarization.jl) when you're updating `graythresh` and `im2bw` 😄

@johnnychen94 
Good point! I will contribute to the original JuliaImages if the function was missing. And if the same functionality exists in other packages, we will provide the API here. But as you are saying providing from living Julia packages is a better way (as the real implementation may be improved).

The goal of this repo is to provide an all in one package, and also, to simplify doing tasks by providing APIs for users that are used to Matlab's syntax.

**Comment by aminya** [comment]
Thank you for help 😃 

I will look into that ImageBinarization repo.

> 
> > replacing raw(img) simply with rawview(img) does not work because of no matching methods
> 
> ```julia
> help?> rawview
> search: rawview
> 
>   rawview(img::AbstractArray{FixedPoint})
> 
> 
>   returns a "view" of img where the values are interpreted in terms of their raw underlying storage. For example, if
>   img is an Array{N0f8}, the view will act like an Array{UInt8}.
> ```
> 
> so `rawview(channelview(img))` would work
> 
> I'm not sure if I understand it correctly, but if `imagetools.jl` is treated as another API specification module, I think there are fewer merits to return/use a "raw" numerical array by default (although MATLAB does it this way) compared to `AbstractArray{<:Colorant}`.

`Imagetools.jl` is a subpackage (like Matlab's toolbox).

You are right about returning. I should keep the functionality with other functions or to add an option to return both types or

**Comment by aminya** [comment]
These are all the toolboxes we plan to add[ MATLAB-to-Julia Functions Mapping](https://en.wikibooks.org/wiki/Julia_for_MATLAB_Users/Index). As you see, a lot of the pages like image processing do not exist. Hopefully, by getting more help from users, we can make this package a good all in one wrapper for Matlab's users. This will draw a lot of users into Julia.

**Comment by johnnychen94** [comment]
The idea is great. However, as a comprehensive MATLAB-to-Julia API specification package, I see two issues here:

* This requires helps from contributors from different backgrounds
* Loading all the packages would be extremely time-consuming

With the usage of [Reexport](https://github.com/simonster/Reexport.jl) and [Requires](https://github.com/MikeInnes/Requires.jl), we can conditionally load and reexport these submodules if their Julia packages are already loaded.

BTW, looks like this repository is inactive and too outdated for quite a long time, perhaps it's better to restart it?

We can also maintain this package in a toolbox-level, e.g., `MatlabImageProcessingToolbox.jl`, and composite them together in `MatlabCompat.jl`

**Comment by aminya** [comment]
> 
> 
> The idea is great. However, as a comprehensive MATLAB-to-Julia API specification package, I see two issues here:
> 
>     * This requires helps from contributors from different backgrounds
> 
>     * Loading all the packages would be extremely time-consuming
> 
> 
> With the usage of [Reexport](https://github.com/simonster/Reexport.jl) and [Requires](https://github.com/MikeInnes/Requires.jl), we can conditionally load and reexport these submodules if their Julia packages are already loaded.
> 
> BTW, looks like this repository is inactive and too outdated for quite a long time, perhaps it's better to restart it?

Agree with that. 👍 

Yes, this repo is old, but I have contacted the owner (@ayakimovich) using email, and he responded once. Hopefully, he will transfer the project to me or I can get write access very soon.

By doing some advertising in the forum probably we can get the help of developers to provide Matlab friendly wrappers here.

**Comment by aminya** [comment]
> We can also maintain this package in a toolbox-level, e.g., `MatlabImageProcessingToolbox.jl`, and composite them together in `MatlabCompat.jl`

This is a better idea. This way we give the choice to the user and also solve the problem for loading time.

**Comment by aminya** [comment]
This pull request is closed and I moved the idea's project to my organization.  https://github.com/juliamatlab

**Pull request closed without merging**